src/prefect/backend/flow_run.py

```diff
@@ -0,0 +1,596 @@
+
+        Args:
+            task: A `prefect.Task` object to use for the lookup. The slug will be
+                pulled from the task to actually perform the query
+            task_slug: A task slug string to use for the lookup
+            task_run_id: A task run uuid to use for the lookup
+
+        Returns:
+            A cached or newly constructed TaskRun instance
+        """
+
+        if task is not None:
+            if not task.slug and not task_slug:
+                raise ValueError(
+                    f"Given task {task} does not have a `slug` set and cannot be "
+                    "used for lookups; this generally occurs when the flow has not "
+                    "been registered or serialized."
+                )
+
+            if task_slug is not None and task_slug != task.slug:
+                raise ValueError(
+                    "Both `task` and `task_slug` were provided but they contain "
+                    "different slug values! "
+                    f"`task.slug == {task.slug!r}` and `task_slug == {task_slug!r}`"
+                )
+
+            # If they've passed a task_slug that matches this task's slug we can
+            # basically ignore the passed `task` object
+            if not task_slug:
+                task_slug = task.slug
+
+        if task_run_id is not None:
+            # Load from the cache if available or query for results
+            result = (
+                self.task_runs[task_run_id]
+                if task_run_id in self._task_runs
+                else TaskRun.from_task_run_id(task_run_id)
+            )
+
+            if task_slug is not None and result.task_slug != task_slug:
+                raise ValueError(
+                    "Both `task_slug` and `task_run_id` were provided but the task "
+                    "found using `task_run_id` has a different slug! "
+                    f"`task_slug == {task_slug!r}` and "
+                    f"`result.task_slug == {result.task_slug!r}`"
+                )
+
+            self._add_task_run(result)
+            return result
+
+        if task_slug is not None:
+
+            if task_slug in self._task_slug_to_task_run_ids:
+                task_run_ids = self._task_slug_to_task_run_ids[task_slug]
+
+                # Check for the 'base' task, for unmapped tasks there should always
+                # just be one run id but for mapped tasks there will be multiple
+                for task_run_id in task_run_ids:
+                    result = self.task_runs[task_run_id]
+                    if result.map_index == -1:
+                        return result
+
+                # We did not find the base mapped task in the cache so we'll
+                # drop through to query for it
+
+            result = TaskRun.from_task_slug(
+                task_slug=task_slug, flow_run_id=self.flow_run_id
+            )
+            self._add_task_run(result)
+            return result
+
+        raise ValueError(
```

Comment:
minor preference to put this at the beginning of the method

Reply:
Hmm that might make more sense now that I added checks for empty `task.slug` up there. It made more sense to fall through at first because I was inferring some values that could still be empty